src/schemes/boundary/open_boundary/system.jl

```diff
@@ -0,0 +1,622 @@
+
+    return hcat(c, edge1, edge2)
+end
+
+@inline function within_boundary_zone(particle_coords, system)
+    (; zone_origin, spanning_set) = system
+    particle_position = particle_coords - zone_origin
+
+    for dim in 1:ndims(system)
+        span_dim = spanning_set[dim]
+        # Checks whether the projection of the particle position
+        # falls within the range of the zone.
+        if !(0 <= dot(particle_position, span_dim) <= dot(span_dim, span_dim))
+
+            # Particle is not in boundary zone.
+            return false
+        end
+    end
+
+    # Particle is in boundary zone.
+    return true
+end
+
+function update_final!(system::OpenBoundarySPHSystem, v, u, v_ode, u_ode, semi, t)
+    if t > 0.0 && !(system.callback_used[])
+        throw(ArgumentError("`UpdateCallback` is required when using `OpenBoundarySPHSystem`"))
+    end
+
+    @trixi_timeit timer() "evaluate characteristics" evaluate_characteristics!(system, v, u,
+                                                                               v_ode, u_ode,
+                                                                               semi, t)
+end
+
+update_open_boundary_eachstep!(system, v_ode, u_ode, semi, t) = system
+
+function update_open_boundary_eachstep!(system::OpenBoundarySPHSystem, v_ode, u_ode,
+                                        semi, t)
+    u = wrap_u(u_ode, system, semi)
+    v = wrap_v(v_ode, system, semi)
+
+    @trixi_timeit timer() "update quantities" update_quantities!(system, v, u, t)
+
+    @trixi_timeit timer() "check domain" check_domain!(system, v, u, v_ode, u_ode, semi)
+
+    @trixi_timeit timer() "update buffer" foreach_system(semi) do system
+        update_system_buffer!(system.buffer)
+    end
+end
+
+# ==== Characteristics
+# J1: Associated with convection and entropy and propagates at flow velocity.
+# J2: Propagates downstream to the local flow
+# J3: Propagates upstream to the local flow
+function evaluate_characteristics!(system, v, u, v_ode, u_ode, semi, t)
+    (; volume, characteristics, previous_characteristics, boundary_zone) = system
+
+    for particle in eachparticle(system)
+        previous_characteristics[1, particle] = characteristics[1, particle]
+        previous_characteristics[2, particle] = characteristics[2, particle]
+        previous_characteristics[3, particle] = characteristics[3, particle]
+    end
+
+    set_zero!(characteristics)
+    set_zero!(volume)
+
+    # Use all other systems for the characteristics
+    foreach_system(semi) do neighbor_system
+        evaluate_characteristics!(system, neighbor_system, v, u, v_ode, u_ode, semi, t)
+    end
+
+    # Only some of the in-/outlet particles are in the influence of the fluid particles.
+    # Thus, we find the characteristics for the particle which are outside the influence
+    # using the average of the values of the previous time step.
+    # Negi (2020) https://doi.org/10.1016/j.cma.2020.113119
```

Comment:
eq. 27 in https://doi.org/10.1016/j.cma.2020.113119

Reply:
Then please add this to the comment.